homeassistant/loader.py

```diff
@@ -1132,43 +1130,39 @@ def bind_hass(func: _CallableT) -> _CallableT:
 
 async def _async_component_dependencies(
     hass: HomeAssistant,
-    start_domain: str,
     integration: Integration,
-    loaded: set[str],
-    loading: set[str],
 ) -> set[str]:
-    """Recursive function to get component dependencies.
+    """Async function to get component dependencies."""
+    loading = set()
+    loaded = set()
 
-    Async friendly.
-    """
-    domain = integration.domain
-    loading.add(domain)
+    async def helper(integration: Integration) -> None:
+        """Recursive async helper to get component dependencies."""
+        domain = integration.domain
+        loading.add(domain)
 
-    for dependency_domain in integration.dependencies:
-        # Check not already loaded
-        if dependency_domain in loaded:
-            continue
+        for dependency_domain in integration.dependencies:
+            dep_integration = await async_get_integration(hass, dependency_domain)
 
-        # If we are already loading it, we have a circular dependency.
-        if dependency_domain in loading:
-            raise CircularDependency(domain, dependency_domain)
+            # If we are already loading it, we have a circular dependency.
+            # We have to check it here, because we need to cover all possible dependants.
+            if conflict := loading.intersection(dep_integration.after_dependencies):
+                raise CircularDependency(conflict.pop(), dependency_domain)
```

Comment:
Can you explain how this check as well as the pop() works?

Reply:
It checks whether there's any overlap between the after_dependencies of the integration and integrations that are already in <loading> (and therefore depend on the current integration). That is a circular dependency.
pop() just gets the first example from the conflicting set.

Reply:
Right. Instead of popping an example, maybe just pass the entire intersection to `CircularDependency`. The type checker will complain, but I don't see a problem with updating it to allow passing sets of strings to it as well as strings.